Google.PowerShell/Provider/GoogleCloudStorageProvider.cs

```diff
@@ -1001,18 +1050,63 @@ private static IEnumerable<Project> ListAllProjects()
+        /// Otherwise, we update the cache and perform the action while doing that (for example,
+        /// we can write the bucket to the command line as they become available instead of writing
+        /// all at once).
+        /// </summary>
+        /// <param name="actionOnBucket"></param>
+        private void PerformActionOnBucketAndOptionallyUpdateCache(Action<Bucket> actionOnBucket)
         {
-            return BucketCache.Value.Values;
+            // If the cache is already initialized and not stale, simply perform the action on each bucket.
+            // Otherwise, we update the cache and perform action on each of the item while doing so.
+            if (!BucketCache.CacheOutOfDate)
+            {
+                Dictionary<string, Bucket> bucketDict = BucketCache.GetLastValueWithoutUpdate();
+                foreach(Bucket bucket in bucketDict.Values)
+                {
+                    actionOnBucket(bucket);
+                }
+            }
+            else
+            {
+                Func<Dictionary<string, Bucket>> functionToUpdateCacheAndPerformActionOnBucket = () => UpdateBucketCacheAndPerformActionOnBucket(actionOnBucket);
+                BucketCache.GetValueWithUpdateFunction(functionToUpdateCacheAndPerformActionOnBucket);
+            }
         }
 
-        private static Dictionary<string, Bucket> UpdateBucketCache()
+        /// <summary>
+        /// Update BucketCache and perform action on each of the bucket while doing so.
+        /// </summary>
+        /// <param name="action"></param>
+        /// <returns></returns>
+        private static Dictionary<string, Bucket> UpdateBucketCacheAndPerformActionOnBucket(Action<Bucket> action)
         {
-            List<Project> projects = ListAllProjects().ToList();
-            // Use ToList() to start all the tasks.
-            List<Task<IEnumerable<Bucket>>> tasks = projects.Select(ListBucketsAsync).ToList();
-            IEnumerable<Bucket> buckets = tasks.SelectMany(task => task.Result);
-            return buckets.ToDictionary(bucket => bucket.Name);
+            BlockingCollection<Bucket> bucketCollections = new BlockingCollection<Bucket>();
+            ConcurrentDictionary<string, Bucket> bucketDict = new ConcurrentDictionary<string, Bucket>();
+            IEnumerable<Project> projects = ListAllProjects();
+            // In each of these tasks, the buckets will be added to the blocking collection bucketCollections.
+            IEnumerable<Task> taskWithActions = projects.Select(project => ListBucketsAsync(project, bucketCollections));
+            // Once all the tasks are done, we signal to the blocking collection that there is nothing to be added.
+            Task.Factory.ContinueWhenAll(taskWithActions.ToArray(), result => { bucketCollections.CompleteAdding(); });
+
+            // bucketCollections.IsCompleted is true if CompleteAdding is called.
+            while (!bucketCollections.IsCompleted)
```

Comment:
I think await would have the same problem. If the user see the operation takes too long, the user can do a Ctrl + C. Since we check for Stopping in ListBucketsAsync, we will stop sending out requests.